src/utils/gitlab_updater.rs

```diff
@@ -0,0 +1,435 @@
+                .unwrap()
+        });
+
+        match RE.captures(url) {
+            Some(cap) => {
+                let owner = cap.name("owner").expect("missing group 'owner'").as_str();
+                let repo = cap.name("repo").expect("missing group 'repo'").as_str();
+                Some(Self {
+                    owner,
+                    repo: repo.strip_suffix(".git").unwrap_or(repo),
+                })
+            }
+            None => None,
+        }
+    }
+}
+
+#[derive(Debug, failure::Fail)]
+#[fail(display = "rate limit reached")]
+struct RateLimitReached;
+
+#[derive(Debug, Deserialize)]
+struct GraphResponse<T> {
+    data: Option<T>,
+    #[serde(default)]
+    errors: Vec<GraphError>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphError {
+    message: String,
+    locations: Vec<GraphErrorLocation>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphErrorLocation {
+    line: u32,
+    column: u32,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphRateLimit {
+    remaining: u32,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GraphNodes<T> {
+    nodes: Vec<T>,
+    rate_limit: GraphRateLimit,
+}
+
+#[derive(Debug, Deserialize)]
+struct GraphProjectNode {
+    project: Option<GraphProject>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GraphProject {
+    id: String,
+    name_with_namespace: String,
+    last_activity_at: Option<DateTime<Utc>>,
+    description: Option<String>,
+    star_count: i64,
+    forks_count: i64,
+    open_issues_count: i64,
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_repository_name() {
+        macro_rules! assert_name {
+            ($url:expr => ($owner:expr, $repo: expr)) => {
+                assert_eq!(
+                    RepositoryName::from_url($url),
+                    Some(RepositoryName {
+                        owner: $owner,
+                        repo: $repo
+                    })
+                );
+            };
+        }
+
+        assert_name!("https://gitlab.com/onur/cratesfyi" => ("onur", "cratesfyi"));
+        assert_name!("http://gitlab.com/onur/cratesfyi" => ("onur", "cratesfyi"));
+        assert_name!("https://www.gitlab.com/onur/cratesfyi" => ("onur", "cratesfyi"));
+        assert_name!("http://www.gitlab.com/onur/cratesfyi" => ("onur", "cratesfyi"));
+        assert_name!("https://gitlab.com/onur/cratesfyi.git" => ("onur", "cratesfyi"));
+        assert_name!("https://gitlab.com/docopt/docopt.rs" => ("docopt", "docopt.rs"));
+        assert_name!("https://gitlab.com/onur23cmD_M_R_L_/crates_fy-i" => (
+            "onur23cmD_M_R_L_", "crates_fy-i"
+        ));
+        assert_name!("https://freedesktop.gitlab.org/test/test" => (
+            "test", "test"
+        ));
+        assert_name!("https://gitlab.freedesktop.org/test/test" => (
+            "test", "test"
+        ));
```

Comment:
We must only support gitlab.com, not other GitLab instances. Otherwise someone could publish a crate with an URL that looks like a GitLab instance and extract our API token.

Reply:
Maybe we should provide a whitelist for that? For example, `gitlab.freedesktop.org` is GNOME gitlab instance. I'll write that.

Reply:
We'd need a per-instance token, `gitlab.com` tokens won't work on `gitlab.freedesktop.org`.

Reply:
I kinda hoped it would but I guess not... Any suggestion for the token naming maybe?

Reply:
Does gitlab allow gathering this info without an API token? That would be ideal - have a token only for gitlab.com, and fetch the info unauthenticated for other instances. I expect we'll have very little traffic for self-hosted instances, I don't think we'll hit rate limits.

Reply:
By default it's 5 requests per minute. So I think it should be fine indeed. I'll check if it works without tokens.

Reply:
Ok, so I can already tell this: the token isn't required (even though it's supposed to limit to 5 requests per minute). However, I wasn't able to trigger the rate limit for some reason...